### PR TITLE
1912 tags endpoint

### PR DESCRIPTION
**Pull request by maskarb**
#1912

Adds the `tags/{infra}/{key}/` endpoint.
Since I was already modifying the urls.py file, I switched out `url()` for `path()` because `url()` may be deprecated in the future ([link](https://docs.djangoproject.com/en/2.2/ref/urls/#url)).

Testing:
1. load some tagged test data.
2. Go to `tags/{infra}/` and take one of the `key`s.
3. Take the key, and go to `tags/{infra}/{key}/.
4. See a populated list of values associated with that key.
5. Apply random filters/group_bys/order_bys. See that the all work.
6. Apply key_only=True, see Bad Request.
7. Try to hit an endpoint for a tag that does not exist, see 404 not found.

results:
```
GET /api/cost-management/v1/tags/aws/

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/?limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/?limit=100&offset=0"
    },
    "data": [
        {
            "key": "version",
            "values": [
                "prod",
                "master",
                "gamma",
                "beta"
            ]
        },
...
```

```
GET /api/cost-management/v1/tags/aws/version/

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/version/?limit=100&offset=0"
    },
    "data": [
        "prod",
        "master",
        "gamma",
        "beta"
    ]
}
```
If using the default Nise data, apply: `?filter[account]=8`
```
GET /api/cost-management/v1/tags/aws/version/?filter[account]=8

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 0,
        "filter": {
            "account": [
                "8"
            ],
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "key_only": false,
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?filter%5Baccount%5D=8&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/version/?filter%5Baccount%5D=8&limit=100&offset=0"
    },
    "data": []
}
```
Apply non-sense:
``` 
GET /api/cost-management/v1/tags/aws/non-sense-tag-which-does-not-exist/

HTTP 404 Not Found
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "errors": [
        {
            "detail": "Not found.",
            "source": "detail",
            "status": 404
        }
    ]
}
```
Try to apply the `key_only` tag query param:
```
GET /api/cost-management/v1/tags/aws/version/?key_only=True
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "errors": [
        {
            "detail": "Invalid query parameter 'key_only'.",
            "source": "details",
            "status": 400
        }
    ]
}
```

### CODE REVIEW

File: koku/api/tags/queries.py [hunk 1]
```diff
@@ -80,6 +80,14 @@ def __init__(self, parameters):
         self._set_start_and_end_dates()
         # super() needs to be called before calling _get_filter()
         self.query_filter = self._get_filter()
+        if parameters.kwargs.get("key"):
+            self.key = parameters.kwargs.get("key")
+            self.query_filter = self._get_key_filter()
+
+    def _get_key_filter(self):
```
Comment (maskarb) [inline koku/api/tags/queries.py:87]:
```suggestion
    def _get_key_filter(self):
        """Add new `exact` QueryFilter that filters on the key name."""
```

File: koku/api/query_params.py [hunk 2]
```diff
@@ -379,7 +368,7 @@ def get_replacement_result(param_res_list, access_list, raise_exception=True):
     """Adjust param list based on access list."""
     if ReportQueryHandler.has_wildcard(param_res_list):
         return access_list
-    if not access_list and not raise_exception:
+    if not (access_list or raise_exception):
```
Comment (blentz) [inline koku/api/query_params.py:371]: Is flipping the `and` to an `or` fixing a bug or creating one? :smiley:

Comment (maskarb) [reply koku/api/query_params.py:371]: It does neither...
```python
>>> not True and not False
False
>>> not True and not True
False
>>> not False and not False
True
>>> not (True or False)
False
>>> not (True or True)
False
>>> not (False or False)
True
```
That change came from this extension: [https://sourcery.ai/](https://sourcery.ai/)

**Comment by maskarb** [comment]
pagination:

```
GET /api/cost-management/v1/tags/aws/version/?limit=1
HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "limit": 1,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=0",
        "next": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=1",
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=3"
    },
    "data": [
        "prod"
    ]
}
```
next page:
```
GET /api/cost-management/v1/tags/aws/version/?limit=1&offset=1
HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: HTTP_X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 4,
        "key_only": false,
        "limit": 1,
        "offset": 1,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
        "group_by": {},
        "order_by": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=0",
        "next": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=2",
        "previous": "/api/cost-management/v1/tags/aws/version/?limit=1",
        "last": "/api/cost-management/v1/tags/aws/version/?limit=1&offset=3"
    },
    "data": [
        "master"
    ]
}
```

**Pull request merged**